PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/configuration_classes/configuration_space.cpp

```diff
@@ -0,0 +1,152 @@
+
+    pushRegister(0, Expansion_ROM_Base_Address, READ_WRITE, 0, 0x000007FE);
+
+    pushRegister(0, Capabilities_Pointer, READ_ONLY, 0, 0);
+}
+
+/**
+ * @brief Making the Only available Configuration Space (Singleton Class)
+ * 
+ * @return shared_ptr<ConfigurationSpace> -> a pointer to our only one Configuration Space
+ */
+ConfigurationSpace * ConfigurationSpace::constructConfigurationSpace()
+{
+    if(configurationSpace == nullptr)
+        configurationSpace = new ConfigurationSpace();
+
+    return configurationSpace;
+}
+
+/**
+    * @brief ConfigurationSpace desructor for deleting all nodes in the linked list to avoid memory leaks
+*/
+ConfigurationSpace::~ConfigurationSpace()
+{
+    Register * current = head_;
+    Register* next;
+
+    while (current)
+    {
+        size_--;
+
+        next = current->getRegisterNext();
+
+        delete current;
+
+        current = next;
+    }
+}
+
+/**
+ * @brief This function is for accepting a visit from a configuration read or write visitor to either read from the configuration space
+ *        or write to it
+ * 
+ * @param visitor -> either a configuration read or write visitor
+ * @return unsigned int -> the value read from the configuration space or zero in case of writing to the configuration space
+ */
+unsigned int ConfigurationSpace::accept(shared_ptr<ConfigurationVisitor> visitor)
+{
+    //return visitor->visitConfigurationSpace(this);
+    return 0;
+}
+
+/**
+ * @brief Getting the size of the linked list (Configuration Space)
+ * 
+ * @return int -> size of the linked list
+ */
+int ConfigurationSpace::getNumberOfRegisters()
+{
+    return size_;
+}
+
+/**
+ * @brief Setting the Master Abort bit in the Status Register when we receive an Unsupported Request
+ *        so a completion with UR will be sent to the requester, and this bit will be set to 1
+ */
+void ConfigurationSpace::setReceivedMasterAbortBit()
+{
+    Register * status = head_;
+    status = status->getRegisterNext()->getRegisterNext()->getRegisterNext();
+
+    status->setRegisterValue((status->getRegisterValue() & 0xFFFFDFFF) | 0x00002000);
+}
+
+/**
+ * @brief This function is be used by the memory class to know whether it can serve any access (read or write) request or not
+ * 
+ * @return int -> whether the Memory Space is enabled or not
+ */
+int ConfigurationSpace::isMemorySpaceEnabled()
+{
+    // Traverse the linked list to find the Command Register
+    Register * current = head_;
+    Register * command = head_;
+    command = command->getRegisterNext()->getRegisterNext();
+
+    // If the Memory Space Enable bit (bit number 2) is set to 1, then return 1, otherwise return 0
+    return (command->getRegisterValue() & 0x00000002);
+}
+
+/**
+ * @brief This function is be used by the IO class to know whether it can serve any access (read or write) request or not
+ * 
+ * @return int -> whether the IO Space is enabled or not
+ */
+int ConfigurationSpace::isIOSpaceEnabled()
+{
```

Comment:
I think we will not work with IO space at all right? 
So why we have this here?

Reply:
I know we don't deal with IO (till now), but it's a part of the configuration space, so you may not implement this part, but it's okay to have it even if we don't use it.

Reply:
okay 👍